### PR TITLE
feat(judicial-system): InvestigationRequest Prosecutors

### PR DESCRIPTION
**Pull request by oddsson**
Prosecutor flow for R-Cases

https://app.asana.com/0/1199153462262248/1200164237014735
https://app.asana.com/0/1199153462262248/1200414912242011

## What

☝️ 
Not included:
- PDFs
- Notifications
- Autofill for prosecutor demands
- Multiple defendants

## Why

This is an important step in supporting "Rannsóknarheimildir" in the justice system.

## Screenshots / Gifs

???

## Checklist:

- [x] I have performed a self-review of my own code
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] Formatting passes locally with my changes
- [x] I have rebased against main before asking for a review

### CODE REVIEW

File: apps/judicial-system/backend/migrations/20210608081018-update-case.js [hunk 1]
```diff
@@ -0,0 +1,168 @@
...
+                    subRes[0][0].length === 1 ? subRes[0][0][0].name : '',
+                  requestedValidToDate: c.requested_valid_to_date,
+                  demands: c.demands,
+                  isolation: c.requested_custody_restrictions?.includes(
+                    'ISOLATION',
+                  ),
+                  isExtension: subRes[1][0].length === 1,
+                  previousDecision:
+                    subRes[1][1].rowCount === 1
+                      ? subRes[1][0][0].decision
```
Comment (oddsson) [inline apps/judicial-system/backend/migrations/20210608081018-update-case.js:88]: 😳

Comment (oddsson) [reply apps/judicial-system/backend/migrations/20210608081018-update-case.js:88]: I'm just going to take your word for this 🙈

File: apps/judicial-system/backend/src/app/modules/case/models/case.model.ts [hunk 2]
```diff
@@ -40,21 +43,43 @@ export class Case extends Model<Case> {
...
+   **********/
+  @Column({
+    type: DataType.STRING,
+    allowNull: true,
+  })
+  @ApiProperty()
+  description: string
+
+  /**********
+   * The case state - exaample: DRAFT
```
Comment (oddsson) [inline apps/judicial-system/backend/src/app/modules/case/models/case.model.ts:81]:
```suggestion
   * The case state - example: DRAFT
```

File: apps/judicial-system/backend/src/app/modules/case/models/case.model.ts [hunk 3]
```diff
@@ -64,122 +89,189 @@ export class Case extends Model<Case> {
...
+   **********/
   @Column({
     type: DataType.ENUM,
     allowNull: true,
     values: Object.values(CaseGender),
   })
   accusedGender: CaseGender
 
+  /**********
+   * The name of the accused´s defender - optional
```
Comment (oddsson) [inline apps/judicial-system/backend/src/app/modules/case/models/case.model.ts:143]:
```suggestion
   * The name of the accused's defender - optional
```

File: apps/judicial-system/backend/src/app/modules/case/models/case.model.ts [hunk 4]
```diff
@@ -64,122 +89,189 @@ export class Case extends Model<Case> {
...
+   **********/
   @Column({
     type: DataType.STRING,
     allowNull: true,
   })
   @ApiProperty()
   defenderName: string
 
+  /**********
+   * The email address of the accused´s defender - optional
```
Comment (oddsson) [inline apps/judicial-system/backend/src/app/modules/case/models/case.model.ts:153]:
```suggestion
   * The email address of the accused's defender - optional
```

File: apps/judicial-system/backend/src/app/modules/case/models/case.model.ts [hunk 5]
```diff
@@ -64,122 +89,189 @@ export class Case extends Model<Case> {
...
+   **********/
   @Column({
     type: DataType.DATE,
     allowNull: true,
   })
   @ApiProperty()
-  requestedCustodyEndDate: Date
+  requestedValidToDate: Date
 
+  /**********
```
Comment (oddsson) [inline apps/judicial-system/backend/src/app/modules/case/models/case.model.ts:242]: I think we should either finish filling these in or remove the empty comments

File: apps/judicial-system/web/src/routes/Court/CourtRecord/CourtRecord.spec.tsx [hunk 6]
```diff
@@ -29,12 +29,15 @@ describe('/domari-krafa/thingbok', () => {
           ...mockJudgeQuery,
           ...mockUpdateCaseMutation(
             [
+              {
```
Comment (oddsson) [inline apps/judicial-system/web/src/routes/Court/CourtRecord/CourtRecord.spec.tsx:32]: Why is this needed?

File: apps/judicial-system/backend/migrations/20210608081018-update-case.js (hunk 1 above)
Comment (gudjong) [reply apps/judicial-system/backend/migrations/20210608081018-update-case.js:88]: Right, there was actually an error there. We cannot easily access rowCount so it is better to check the length of the result. I forgot to change that in one place.

File: apps/judicial-system/backend/src/app/modules/case/models/case.model.ts (hunk 2 above)
Comment (gudjong) [reply apps/judicial-system/backend/src/app/modules/case/models/case.model.ts:81]: My keyboard 😱

File: apps/judicial-system/backend/src/app/modules/case/models/case.model.ts (hunk 5 above)
Comment (gudjong) [reply apps/judicial-system/backend/src/app/modules/case/models/case.model.ts:242]: I will fill in the comments as I think they are very important and increasingly so.

File: apps/judicial-system/web/pages/krafa/domkrofur-og-lagaakvaedi/[id].tsx [hunk 7]
```diff
@@ -1,3 +1,3 @@
-import { StepThree } from '@island.is/judicial-system-web/src/routes/Prosecutor/CreateDetentionRequest/StepThree/StepThree'
+import { StepThree } from '@island.is/judicial-system-web/src/routes/Prosecutor/CustodyRequest/StepThree/StepThree'
```
Comment (gudjong) [inline apps/judicial-system/web/pages/krafa/domkrofur-og-lagaakvaedi/[id].tsx:1]: This path is also used the travel bans, so perhaps we should change to a name that reflects that. Don't know what that should be though.

File: apps/judicial-system/web/src/routes/Court/CourtRecord/CourtRecord.spec.tsx (hunk 6 above)
Comment (gudjong) [reply apps/judicial-system/web/src/routes/Court/CourtRecord/CourtRecord.spec.tsx:32]: Because prosecutorDemands is autofilled from demands.

The logic goes like this:
1.  The prosecutor fills in requestedValidToDate
2. In the prosecutor flow, the client autfills demands using formatPresecutorDemands
3. In the court flow, the client autofills prosecutorDemands using demands

Comment (gudjong) [reply apps/judicial-system/web/src/routes/Court/CourtRecord/CourtRecord.spec.tsx:32]: I guess we no longer need the second prosecutorDemands entry. Will fix.

**Review by gudjong — APPROVED** [review]
Looks good after various updates. Did not get all the details in the new client r-case part, but that will be tester before release unhidden to prod. I review modifications to old code carefully.
